File: C++/Stacks & Queues/28_interleaving_2_halves.cpp
/*
Idea:
inserting all elements in a vector & using 2 pointers, one from start and one from mid, appending the answer
*/

//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution {
public:
    // Function to rearrange the elements of the queue
    vector<int> rearrangeQueue(queue<int>& q) {
        // Vector to store the rearranged elements
        vector<int> ans, temp;

        // Step 1: Move elements from the queue to a temporary vector
        while (!q.empty()) {
            temp.push_back(q.front());  // Store the front element of the queue in the temporary vector
            q.pop();  // Remove the front element from the queue
        }

        // Pointers i and j for traversing the temporary vector
        int i = 0, j = temp.size() / 2;

        // Step 2: Rearrange the elements in the temporary vector and store in the result vector
        while (j < temp.size()) {
            ans.push_back(temp[i++]);  // Append the i-th element to the result vector
            ans.push_back(temp[j++]);  // Append the j-th element to the result vector
        }

        // Return the rearranged vector
        return ans;
    }
};

//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;

        queue<int> q;
        for (int i = 0;i < n;i++) {
            int a;
            cin >> a;
            q.push(a);
        }
        Solution ob;
        vector<int> ans = ob.rearrangeQueue(q);
        for (int i = 0;i < ans.size();i++) {
            cout << ans[i] << " ";
        }
        cout << endl;
    }

    return 0;
}
// } Driver Code Ends

File: Python/Stacks & Queues/28_interleaving_2_halves.py
"""
Idea:
inserting all elements in a vector & using 2 pointers, one from start and one from mid, appending the answer
"""

from typing import List


class Solution:
    def rearrangeQueue(self, N: int, q: List[int]) -> List[int]:
        # List to store the rearranged elements
        ans, temp = [], []

        # Step 1: Move elements from the queue to a temporary list
        for elem in q:
            temp.append(elem)  # Get and append the front element of the queue to the temporary list

        # Pointers i and j for traversing the temporary list
        i, j = 0, len(temp) // 2

        # Step 2: Rearrange the elements in the temporary list and store in the result list
        while j < len(temp):
            ans.append(temp[i])  # Append the i-th element to the result list
            ans.append(temp[j])  # Append the j-th element to the result list
            i += 1
            j += 1

        # Return the rearranged list
        return ans


# {
# Driver Code Starts

class IntArray:
    def __init__(self) -> None:
        pass

    def Input(self, n):
        arr = [int(i) for i in input().strip().split()]  # array input
        return arr

    def Print(self, arr):
        for i in arr:
            print(i, end=" ")
        print()


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        N = int(input())

        q = IntArray().Input(N)

        obj = Solution()
        res = obj.rearrangeQueue(N, q)

        IntArray().Print(res)

# } Driver Code Ends